### PR TITLE
Reduce telemetry RAM footprint and improve usability/readability

### PR DESCRIPTION
**Pull request by konstantin-msft**
RAM footprint reduction:
- Reduce `eventsByCorrelationId` map RAM usage by ~88%. We store 8 sub-measurements on average as separate events in `eventsByCorrelationId` map that are rolled-up into root event on ending.
- Reduce `measurementsByCorrelationId` map RAM usage by 100%. Measurements are not stored in map anymore. 

Code changes:
- Store a single root event per correlation id. Sub-measurements are populated to root event on ending.
- Store static fields and counters in root event instead of caching them.
- Don't cache measurements. Measurement is an attribute of instance of `InProgressPerformanceEvent`. 
- Deprecate `flushMeasurement` event and move all logic to `endMeasurement`. This allows to end measurement with a single method.
- Add a method to flush incomplete browser measurements.

Notes:
- `PerformanceClient.flushMeasurement()` will be removed completely right after i remove all its` references in 1p repo.

### CODE REVIEW

File: lib/msal-common/src/telemetry/performance/IPerformanceClient.ts [hunk 1]
```diff
@@ -21,6 +21,7 @@ export type InProgressPerformanceEvent = {
 export interface IPerformanceClient {
     startMeasurement(measureName: PerformanceEvents, correlationId?: string): InProgressPerformanceEvent;
     endMeasurement(event: PerformanceEvent): PerformanceEvent | null;
+    // @deprecated use "endMeasurement" instead
```
Comment (peterzenz) [inline lib/msal-common/src/telemetry/performance/IPerformanceClient.ts:24]: This is OK for a patch version?

Comment (peterzenz) [reply lib/msal-common/src/telemetry/performance/IPerformanceClient.ts:24]: Looks like it isn't going away so this is likely OK

**Review by peterzenz — APPROVED** [review]

File: lib/msal-common/src/telemetry/performance/IPerformanceClient.ts (hunk 1 above)
Comment (konstantin-msft) [reply lib/msal-common/src/telemetry/performance/IPerformanceClient.ts:24]: I will mark a consequent PR that removes `flushMeasurement` method as minor.

File: lib/msal-browser/src/app/ClientApplication.ts [hunk 2]
```diff
@@ -561,15 +553,11 @@ export abstract class ClientApplication {
                                 subErrorCode: error.subError,
                                 success: false
                             });
-                            atbcMeasurement.flushMeasurement();
                             throw error;
                         });
                     this.hybridAuthCodeResponses.set(hybridAuthCode, response);
                 } else {
                     this.logger.verbose("Existing acquireTokenByCode request found", request.correlationId);
-                    atbcMeasurement.endMeasurement({
```
Comment (tnorling) [inline lib/msal-browser/src/app/ClientApplication.ts:570]: Why was this removed?

File: lib/msal-browser/src/app/PublicClientApplication.ts [hunk 3]
```diff
@@ -157,16 +156,12 @@ export class PublicClientApplication extends ClientApplication implements IPubli
...
                         success: false
                     });
-                    atsMeasurement.flushMeasurement();
                     throw error;
                 });
             this.activeSilentTokenRequests.set(silentRequestKey, response);
             return response;
         } else {
             this.logger.verbose("acquireTokenSilent has been called previously, returning the result from the first call", correlationId);
-            atsMeasurement.endMeasurement({
```
Comment (tnorling) [inline lib/msal-browser/src/app/PublicClientApplication.ts:167]: Why was this removed?

File: lib/msal-common/src/telemetry/performance/PerformanceClient.ts [hunk 4]
```diff
@@ -269,34 +243,27 @@ export abstract class PerformanceClient implements IPerformanceClient {
...
         // Store in progress events so they can be discarded if not ended properly
         this.cacheEventByCorrelationId(inProgressEvent);
 
-        const staticFields: StaticFields = {
-            appName: this.applicationTelemetry?.appName,
-            appVersion: this.applicationTelemetry?.appVersion,
-        };
-        this.addStaticFields(staticFields, eventCorrelationId);
-        this.cacheMeasurement(inProgressEvent, validMeasurement);
+        // this.cacheMeasurement(inProgressEvent, validMeasurement);
```
Comment (tnorling) [inline lib/msal-common/src/telemetry/performance/PerformanceClient.ts]: Should this be uncommented or removed?

File: lib/msal-common/src/telemetry/performance/PerformanceClient.ts [hunk 5]
```diff
@@ -27,30 +32,10 @@ export abstract class PerformanceClient implements IPerformanceClient {
...
-     * @type {Map<string, Map<string, PerformanceEvent>>}
-     */
-    protected eventsByCorrelationId: Map<string, Map<string, PerformanceEvent>>;
-
-    /**
-     * Fields to be emitted which are scoped to the top level request and whose value will not change in submeasurements
-     * For example: App name, version, etc.
+     * @type {Map<string, PerformanceEvent>}
      */
-    protected staticFieldsByCorrelationId: Map<string, StaticFields>;
```
Comment (tnorling) [inline lib/msal-common/src/telemetry/performance/PerformanceClient.ts:40]: Removing these is a breaking change. This should be a major bump of common

File: lib/msal-common/src/telemetry/performance/PerformanceClient.ts [hunk 6]
```diff
@@ -318,51 +285,80 @@ export abstract class PerformanceClient implements IPerformanceClient {
...
      * @param fields
      * @param correlationId
      */
-    addStaticFields(fields: StaticFields, correlationId: string) : void{
-        const existingStaticFields = this.staticFieldsByCorrelationId.get(correlationId);
-        if (existingStaticFields) {
-            this.logger.trace("PerformanceClient: Updating static fields");
-            this.staticFieldsByCorrelationId.set(correlationId, {...existingStaticFields, ...fields});
+    addStaticFields(fields: StaticFields, correlationId: string) : void {
+        this.logger.trace("PerformanceClient: Upserting static fields");
```
Comment (tnorling) [inline lib/msal-common/src/telemetry/performance/PerformanceClient.ts]: typo:
```suggestion
        this.logger.trace("PerformanceClient: Updating static fields");
```

File: lib/msal-browser/src/app/ClientApplication.ts (hunk 2 above)
Comment (konstantin-msft) [reply lib/msal-browser/src/app/ClientApplication.ts:570]: We don't need to end measurements before discarding them because no events are emitted in that case.

File: lib/msal-browser/src/app/PublicClientApplication.ts (hunk 3 above)
Comment (konstantin-msft) [reply lib/msal-browser/src/app/PublicClientApplication.ts:167]: We don't need to end measurements before discarding them because no events are emitted in that case.

File: lib/msal-common/src/telemetry/performance/PerformanceClient.ts (hunk 5 above)
Comment (konstantin-msft) [reply lib/msal-common/src/telemetry/performance/PerformanceClient.ts:40]: I would rather disagree because:
1. It's does not change API signature
2. We don't have any classes inheriting base `PerformanceClient` class that access `staticFieldsByCorrelationId` directly.

File: lib/msal-common/src/telemetry/performance/PerformanceClient.ts (hunk 4 above)
Comment (konstantin-msft) [reply lib/msal-common/src/telemetry/performance/PerformanceClient.ts]: fixed

File: lib/msal-common/src/telemetry/performance/PerformanceClient.ts (hunk 6 above)
Comment (konstantin-msft) [reply lib/msal-common/src/telemetry/performance/PerformanceClient.ts]: fixed

File: lib/msal-common/src/telemetry/performance/PerformanceClient.ts (hunk 5 above)
Comment (tnorling) [reply lib/msal-common/src/telemetry/performance/PerformanceClient.ts:40]: `PerformanceClient` is exported from common therefore removing anything public or protected is breaking. It's not a big deal though, we just need to mark this as a major change in the msal-common changefile. Please do make sure anything that is not used outside of this class is marked `private` to avoid this churn in the future though.

Comment (konstantin-msft) [reply lib/msal-common/src/telemetry/performance/PerformanceClient.ts:40]: Sounds good. Changed to major version bump.

**Review by tnorling — APPROVED** [review]
